orca/service_impl.go

```diff
@@ -0,0 +1,173 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package orca
+
+import (
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/status"
+
+	v3orcaservicegrpc "github.com/cncf/xds/go/xds/service/orca/v3"
+	v3orcaservicepb "github.com/cncf/xds/go/xds/service/orca/v3"
+)
+
+// Server provides an implementation of the OpenRcaService as defined in the
+// [ORCA] service protos.
+//
+// Server applications can use the SetXxx() methods to record measurements
+// corresponding to backend metrics, which eventually get pushed to clients who
+// have initiated the SteamCoreMetrics streaming RPC.
+//
+// [ORCA]: https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto
+type Server struct {
+	v3orcaservicegrpc.UnimplementedOpenRcaServiceServer
+
+	minReportingInterval time.Duration
+	recorder             *metricRecorder
+}
+
+// ServerOptions contains options to configure the ORCA server implementation.
+type ServerOptions struct {
+	// MinReportingInterval sets the lower bound for how often out-of-band
+	// metrics are reported on the streaming RPC initiated by the client.  If
+	// unspecified or less than the default value of 30s, the default is used.
+	// Clients may request a higher value as part of the StreamCoreMetrics
+	// streaming RPC.
+	MinReportingInterval time.Duration
+}
+
+// Register creates a new ORCA service implementation configured using the
+// provided options and registers the same on the provided service registrar.
+func Register(s grpc.ServiceRegistrar, opts ServerOptions) (*Server, error) {
+	// TODO(easwars): Once the generated pb.gos are updated, we wont need this
+	// type assertion any more, since the new versions accept a
+	// grpc.ServiceRegistrar in their registration functions.
+	srv, ok := s.(*grpc.Server)
+	if !ok {
+		return nil, fmt.Errorf("concrete type of provided grpc.ServiceRegistrar is %T, only supported type is %T", s, &grpc.Server{})
+	}
+	server := NewServer(opts)
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(srv, server)
+	return server, nil
+}
+
+// NewServer returns a new ORCA service implementation configured using the
+// provided options. This can be used to customize the registration of the
+// returned ORCA server. Most usages should prefer to use Register instead.
+func NewServer(opts ServerOptions) *Server {
+	if opts.MinReportingInterval < internal.MinORCAReportingInterval {
+		opts.MinReportingInterval = internal.MinORCAReportingInterval
+	}
+	return &Server{
+		minReportingInterval: opts.MinReportingInterval,
+		recorder:             newMetricRecorder(),
+	}
+}
+
+func (s *Server) determineReportingInterval(req *v3orcaservicegrpc.OrcaLoadReportRequest) time.Duration {
+	if req.GetReportInterval() == nil {
+		return s.minReportingInterval
+	}
+	interval := req.GetReportInterval()
+	if err := interval.CheckValid(); err != nil {
```

Comment:
Why do this extra check?

From what I can tell, there's some definition of "normal" in a time.Duration, but the library deals with it in a pretty sane manner.  Why not just use what it figures out?

> The AsDuration method performs the conversion on a best-effort basis. Durations with denormal values (e.g., nanoseconds beyond -99999999 and +99999999, inclusive; or seconds and nanoseconds with opposite signs) are normalized during the conversion to a time.Duration. To manually check for invalid Duration per the documented limitations in duration.proto, additionally call the CheckValid method:

Reply:
We receive this value on the wire, and I believe this value could have invalid duration value because `durationpb.New()` does not do any kind of validation. https://github.com/protocolbuffers/protobuf-go/blob/v1.28.1/types/known/durationpb/duration.pb.go#L167

The `duration.AsDuration()` implementation https://github.com/protocolbuffers/protobuf-go/blob/v1.28.1/types/known/durationpb/duration.pb.go#L176 returns minInt or maxInt in the case of an overflow.

If we use, `duration.CheckValid()`, we will be able to return a meaningful error to the user and still continue to use the configured minimum.

Why do you see this extra check as problematic? Do you just see it as a wasted effort?

Reply:
GitHub is so messed up. None of the review comments show up along with the source code and I'm not able to reply to them all at once. I have taken care of all comments except this one.

I don't quite understand what you are suggesting here though. Why should we check for `dur > 30*time.Minute`?

There are multiple minimums:
- There is an absolute minimum, which is 30s, and this is what we call the default. 
- There is a minimum enforced by the user when they create the service. This is `ServiceOptions.MinReportingInterval`, and this needs to be greater than or eq to the absolute minimum of 30s. If it is less than the absolute minimum, we end up using the absolute minimum for this.
- Then, the incoming RPC can specify an interval, and this needs to be greater than the configured service minimum. If not, we use the service minimum.

Reply:
Simplified it to match Java's implementation, where the only things that are checked are:
- whether it is specified or not
- whether it is negative
- whether it is less than the configured minimum.